### ISSUE TITLE
[question] What is the best way to create a lockfile for a multi-platform CI/CD pipeline?

### ISSUE THREAD

**Issue by RochaStratovan** [conan-io/conan#15817]
### What is your question?

I'm trying to think of how to use conan with our CI/CD system so that Windows and Linux are using the same dependency version, given that Windows and/or Linux may have a few different package dependencies.

Consider if we had a dependency tree such as:

```bash
Executable-A
   ├───Lib-B (common)
   ├───Lib-C (Windows only)
   └───Lib-D (Linux only)
```

My initial thoughts are that the _Executable-A_ CI/CD pipeline starts with two jobs to create the lockfile that would be used by the rest of the pipeline such as:

1. Job 1: Windows Runner

   1. Create the initial lockfile

      ```bash
      conan install .
      conan lock create .
      ```

   1. Store the lockfile as an artifact

1. Job 2: Linux Runner

   1. Inherits/uses the Windows lockfile artifact

   1. Update the Windows lockfile

      ```bash
      conan install . --lockfile-partial --lockfile-out=conan.lock
      ```

   1. Store the lockfile as an artifact

After this the Windows and Linux build steps would "inherit" lockfile artifact from Linux and then build from one lockfile.

Is there another way to do this? A better way?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @RochaStratovan 

Thanks for your question, this is a good one.

We think that in many cases it will be more robust to capture the lockfile for Windows and Linux quickly at the beginning, in the same machine. Something like:

```bash
$ conan lock create . -pr=windows
$ conan lock create . -pr=linux
```
In most cases it should be possible to do this in the same machine, as this doesn't really install the binaries, and it should also be pretty fast. This allows reducing a lot the potential time in which there could be some divergence in versions.

If you plan also to distribute a build of a dependency graph in multiple agents, you might also want to run ``conan graph build-order`` for the configurations you need to support, also here, in this preparation stage.

Then, when you have the lockfile and you know what do you want to build, then at that point you launch the jobs that do the actual build, and you pass the lockfile and the necessary information to the jobs.

**Comment by RochaStratovan** [comment]
Hi @memsharded,

I thought about the method you are suggesting, using a specific configuration for this. However, I started hitting issues that I didn't know how to resolve.

What would be the proper way to create a profile file?

The example says to do `conan profile detect --force` for the demo but then in the [conan profile](https://docs.conan.io/2/reference/commands/profile.html) documenation it says that it's not recommend to use `conan profile detect`, but instead define our own profile.

I don't see any documentation that explains how I would create my own profile or why I would choose one setting vs another.

Defining our own profile leads to another challenge. Distributing the profile. There are some coments about using `conan config install`.

I found some terse examples for `conan config install`, that look hopeful, but I don't see any example for how to create the installation source.

<br>

What if the profile doesn't match the machine?

I see that the profile compiler configuration identifies the version of the compiler. However, developers are free to upate their version of Visual Studio at their own pace. The CI/CD machines update their version of VS2017, VS2019 and VS2022 once a month. So I'm guessing the centralized configuration would need to be updated at the same time.

What happens if devs are using version 190 of VS2022, while their configuration from `conan config install` clubs it to version 193?

**Comment by memsharded** [comment]
> I thought about the method you are suggesting, using a specific configuration for this. However, I started hitting issues that I didn't know how to resolve.

> What would be the proper way to create a profile file?

Just writing them, putting in your git repo, and sharing them with ``conan config install``.
There isn't any magic about profiles, just the decision of what compilers and configurations you want to use in production. In our experience this isn't and it shouldn't be a non deliberate and explicit decision. Teams decide which tools and configurations they use for their projects, it shouldn't be something random that can be changed arbitrarily by any developer. At least for production environments.

> What if the profile doesn't match the machine?

> I see that the profile compiler configuration identifies the version of the compiler. However, developers are free to upate their version of Visual Studio at their own pace. The CI/CD machines update their version of VS2017, VS2019 and VS2022 once a month. So I'm guessing the centralized configuration would need to be updated at the same time.

You will want to have explicit profiles at least for CI to use the versions you want in CI.
Then, developers auto-detection might work for local development. But if there is any other shared configuration, then developers should be ``conan install ... -pr=vs2019.profile`` if they are using that version. Because there are many other combinations, you might want to force toolset v142 and developers with VS2022 can still use it fine as long as it is configured in the profile (and they  have it installed of course)

In any case, I'd strongly recommend trying to move forward your developers with the toolchain. There are many other aspects, like enabling some C++ standard in your codebase, which becomes very problematic with that policy. Once you enable support and move your CI to the latest VS2022 there is really not much sense to allow devs keep using VS2017. Bugs can be different, reproductions can be different, performance can be different... it introduces unnecessarily variability. I know it is not always easy, but in my opinion the advantages clearly outperform the disadvantages.

Another story is when it is really necessary to support different toolchain versions, because of distributing to other customers, etc. In this cases, the profiles as a mechanism to pin and share configuration is again relevant.

> What if the profile doesn't match the machine?

Well, in the case of VS, depending on the build system, it might be able to detect discrepancies and fail, like "the configuration wanted to use v142, but it cannot be found (because you have only installed v143)". It is true that there are other scenarios, like gcc versions, in which it is not possible to raise errors, and discrepancies might be unnoticed. At the end of the day, it is responsibility of the users, if they are ``conan install ... -pr=gcc9``, it is expected that they will have gcc9 in their system. 
It is also true that these discrepancies, in developers machines, if undetected, they aren't necessarily very problematic always. As long as they cannot upload those packages to the server (and they shouldn't be allowed to do so, only CI can create packages), it can be mostly ok for most cases.

**Comment by RochaStratovan** [comment]
@memsharded,

Thank you for sharing your insights.


> Just writing them, putting in your git repo, and sharing them with conan config install.

Part of my challenge is I don't see documentation or example for the layout/format of the config repository/file that is used with the `conan config` command.

I'm going to devote some time later today to play with this command to see if I can hack it out.


> Another story is when it is really necessary to support different toolchain versions, because of distributing to other customers, etc. In this cases, the profiles as a mechanism to pin and share configuration is again relevant.

This is our situation, my company is often contractually required to provide multiple instances of some our library software. We often have to deliver VS2017, VS2019 and VS2022 -- even though a VS2017 library would work for all of those scenarios due to Microsoft's recent commitment to backwards compatibility.

Moreover, I now realize I was confusing concepts. In the profile information I see items such as the following:

```bash
settings
  arch: x86_64
  build_type: Release
  compiler: msvc
  compiler.cppstd: 14
  compiler.runtime: dynamic
  compiler.runtime_type: Release
  compiler.version: 193           <<<<<<<<<<<<<<
  os: Windows
options
  shared: False
```

I had forgotten that for conan, version is the version of the MSVC (cl.exe) compiler and not the IDE version - for example, I was starting VS2022, going to the about page and seeing 17.8.6. THIS is what the devs have control over. They may not have the latest version of VS20??.

<br>

TL;DR:

I understand you recommend using the sequence

```bash
$ conan lock create . -pr=windows
$ conan lock create . -pr=linux
$ conan lock create . -pr=<other-profile>
```

In one job, at the beginning and then using that lockfile through out the pipeline. Moreover, the `conan config ...` command can be used for sharing/managing common configurations

Thank you for the help.

**Comment by memsharded** [comment]
Thanks for the feedback @RochaStratovan 

> I understand you recommend using the sequence

Yes, I think that is correct, capturing the lockfiles for all configs early in the process is recommended.

> Moreover, the conan config ... command can be used for sharing/managing common configurations

Totally. We are also working to add more features to this ``conan config install`` too
